File: tests/test_doom_specials.py
import unittest

from doom_map import LineDef, Sector, SideDef, Vertex
from doom_specials import (
    Ceiling, Door, FloorMover, LightFlash, Platform, advance_specials,
    trigger_line_special,
)


class DoorRequiredTest(unittest.TestCase):
    """specials.md의 test_required Doom.Door 블록을 그대로 옮긴 테스트."""

    def test_open_door_reaches_target_and_reports_finished(self):
        sector = Sector(0, 0, "F", "C", 160, 0, 0)
        door = Door(sector, "Open", targetCeilingHeight=128, speed=8)
        still_going = True
        for _ in range(1000):
            still_going = door.tick()
            if not still_going:
                break
        self.assertEqual(sector.ceilingHeight, 128)
        self.assertFalse(still_going)

    def test_open_then_close_reopens_if_blocked(self):
        sector = Sector(0, 0, "F", "C", 160, 0, 0)
        door = Door(sector, "OpenThenClose", targetCeilingHeight=128, speed=64)
        door.waitTicsRemaining = 2
        for _ in range(3):
            door.tick()
        self.assertEqual(sector.ceilingHeight, 128)


class FloorMoverTest(unittest.TestCase):
    def test_floor_rises_to_target(self):
        sector = Sector(0, 64, "F", "C", 160, 0, 0)
        mover = FloorMover(sector, targetFloorHeight=32, speed=8)
        while mover.tick():
            pass
        self.assertEqual(sector.floorHeight, 32)


class CeilingCrusherTest(unittest.TestCase):
    def test_repeating_crusher_bounces_between_top_and_bottom(self):
        sector = Sector(0, 128, "F", "C", 160, 0, 0)
        crusher = Ceiling(sector, topHeight=128, bottomHeight=32, speed=32,
                           direction="Down", repeats=True)
        heights = []
        for _ in range(20):
            crusher.tick()
            heights.append(sector.ceilingHeight)
        self.assertIn(32, heights)
        self.assertTrue(crusher.tick())  # 크러셔는 절대 끝나지 않는다(항상 True).


class PlatformTest(unittest.TestCase):
    def test_lift_descends_then_stops(self):
        sector = Sector(64, 128, "F", "C", 160, 0, 0)
        lift = Platform(sector, lowHeight=0, highHeight=64, speed=16, direction="Down")
        while lift.tick():
            pass
        self.assertEqual(sector.floorHeight, 0)


class LightFlashTest(unittest.TestCase):
    def test_low_roll_picks_max_light(self):
        sector = Sector(0, 0, "F", "C", 160, 0, 0)
        flash = LightFlash(sector, maxLight=200, minLight=80, rng=lambda: 0.0, ticsUntilNextChange=1)
        flash.tick()
        self.assertEqual(sector.lightLevel, 200)

    def test_high_roll_picks_min_light(self):
        sector = Sector(0, 0, "F", "C", 160, 0, 0)
        flash = LightFlash(sector, maxLight=200, minLight=80, rng=lambda: 0.9, ticsUntilNextChange=1)
        flash.tick()
        self.assertEqual(sector.lightLevel, 80)


def _make_door_linedef(requiredKey=None):
    v1, v2 = Vertex(0, 0), Vertex(0, 64)
    frontSector = Sector(0, 0, "F", "C", 160, 0, 1)
    backSector = Sector(0, 0, "F", "C", 160, 0, 0)
    front = SideDef(0, 0, "-", "-", "-", frontSector)
    back = SideDef(0, 0, "-", "-", "-", backSector)
    line = LineDef(v1, v2, flags=0, special=1, tag=1, frontSide=front, backSide=back)
    return line, backSector


class LockedDoorTest(unittest.TestCase):
    """specials.md에 새로 추가한 requiredKey/잠긴 문 판정을 검증한다."""

    def test_locked_door_does_not_open_without_matching_key(self):
        line, sector = _make_door_linedef()
        active = []

        class Player:
            keys = []

        def make_mover(l):
            return Door(sector, "Open", targetCeilingHeight=128, requiredKey="blue")

        def is_locked_check(l):
            return Door(sector, "Open", targetCeilingHeight=128, requiredKey="blue")

        mover = trigger_line_special(line, "Use", Player(), active, make_mover, is_locked_check)
        self.assertIsNone(mover)
        self.assertEqual(active, [])
        self.assertNotEqual(line.special, 0)  # 다시 시도할 수 있게 special이 남아 있다.

    def test_locked_door_opens_with_matching_key(self):
        line, sector = _make_door_linedef()
        active = []

        class Player:
            keys = ["blue"]

        def make_mover(l):
            return Door(sector, "Open", targetCeilingHeight=128, requiredKey="blue")

        def is_locked_check(l):
            return Door(sector, "Open", targetCeilingHeight=128, requiredKey="blue")

        mover = trigger_line_special(line, "Use", Player(), active, make_mover, is_locked_check)
        self.assertIsNotNone(mover)
        self.assertEqual(active, [mover])


class SpecialsProgressTest(unittest.TestCase):
    """specials.md의 '## Feature: 특수 효과 진행' 검증."""

    def test_finished_movers_are_removed_still_active_kept(self):
        sectorA = Sector(0, 0, "F", "C", 160, 0, 0)
        sectorB = Sector(0, 0, "F", "C", 160, 0, 0)
        finished_soon = FloorMover(sectorA, targetFloorHeight=1, speed=1)
        long_running = FloorMover(sectorB, targetFloorHeight=100, speed=1)
        active = [finished_soon, long_running]
        active = advance_specials(active)
        self.assertIn(long_running, active)
        # finished_soon은 1 tic만에 목표(1)에 도달해 제거된다.
        self.assertNotIn(finished_soon, active)


if __name__ == "__main__":
    unittest.main()

File: tests/cpp/test_doom_specials.cpp
// examples/doom/specials.md 검증 (C++17).
#include "doom_specials.hpp"
#include "minitest.hpp"
#include <vector>

using namespace Doom;
using namespace Doom::Map;

TEST(door_open_reaches_target_and_reports_finished) {
    Sector sector{0, 0, "F", "C", 160, 0, 0};
    Door door(&sector, "Open", 128, 8);
    bool stillGoing = true;
    for (int i = 0; i < 1000 && stillGoing; i++) stillGoing = door.tick();
    CHECK_EQ(sector.ceilingHeight, 128.0f);
    CHECK(!stillGoing);
}

TEST(door_open_then_close_reopens_if_blocked) {
    Sector sector{0, 0, "F", "C", 160, 0, 0};
    Door door(&sector, "OpenThenClose", 128, 64);
    door.waitTicsRemaining = 2;
    for (int i = 0; i < 3; i++) door.tick();
    CHECK_EQ(sector.ceilingHeight, 128.0f);
}

TEST(floor_rises_to_target) {
    Sector sector{0, 64, "F", "C", 160, 0, 0};
    FloorMover mover(&sector, 32, 8);
    while (mover.tick()) {}
    CHECK_EQ(sector.floorHeight, 32.0f);
}

TEST(ceiling_crusher_repeats_forever) {
    Sector sector{0, 128, "F", "C", 160, 0, 0};
    Ceiling crusher(&sector, 128, 32, 32, "Down", true, true);
    bool sawBottom = false;
    for (int i = 0; i < 20; i++) {
        crusher.tick();
        if (sector.ceilingHeight == 32.0f) sawBottom = true;
    }
    CHECK(sawBottom);
    CHECK(crusher.tick());  // 크러셔는 절대 끝나지 않는다.
}

TEST(platform_lift_descends_then_stops) {
    Sector sector{64, 128, "F", "C", 160, 0, 0};
    Platform lift(&sector, 0, 64, 16, "Down");
    while (lift.tick()) {}
    CHECK_EQ(sector.floorHeight, 0.0f);
}

TEST(light_flash_low_roll_picks_max) {
    Sector sector{0, 0, "F", "C", 160, 0, 0};
    LightFlash flash(&sector, 200, 80, [] { return 0.0; }, 1);
    flash.tick();
    CHECK_EQ(sector.lightLevel, 200);
}

TEST(light_flash_high_roll_picks_min) {
    Sector sector{0, 0, "F", "C", 160, 0, 0};
    LightFlash flash(&sector, 200, 80, [] { return 0.9; }, 1);
    flash.tick();
    CHECK_EQ(sector.lightLevel, 80);
}

struct FakePlayer {
    std::vector<std::string> keys;
};

static LineDef makeDoorLineDef(Sector& front, Sector& back, Vertex& v1, Vertex& v2, SideDef& fSide,
                                SideDef& bSide) {
    fSide.sector = &front;
    bSide.sector = &back;
    return LineDef{&v1, &v2, 0, 1, 1, &fSide, &bSide};
}

TEST(locked_door_does_not_open_without_matching_key) {
    Vertex v1{0, 0}, v2{0, 64};
    Sector front{0, 0, "F", "C", 160, 0, 1}, back{0, 0, "F", "C", 160, 0, 0};
    SideDef fSide{0, 0, "-", "-", "-", nullptr}, bSide{0, 0, "-", "-", "-", nullptr};
    LineDef line = makeDoorLineDef(front, back, v1, v2, fSide, bSide);

    std::vector<std::shared_ptr<SectorMover>> active;
    FakePlayer player;  // 키 없음.

    auto makeMover = [&](LineDef*) {
        return std::make_shared<Door>(&back, "Open", 128.0f, 4.0f, std::optional<std::string>("blue"));
    };
    auto isLockedCheck = [&](LineDef*) {
        return std::make_shared<Door>(&back, "Open", 128.0f, 4.0f, std::optional<std::string>("blue"));
    };

    auto mover = triggerLineSpecial(&line, "Use", &player, active, makeMover, isLockedCheck);
    CHECK(mover == nullptr);
    CHECK(active.empty());
    CHECK(line.special != 0);
}

TEST(locked_door_opens_with_matching_key) {
    Vertex v1{0, 0}, v2{0, 64};
    Sector front{0, 0, "F", "C", 160, 0, 1}, back{0, 0, "F", "C", 160, 0, 0};
    SideDef fSide{0, 0, "-", "-", "-", nullptr}, bSide{0, 0, "-", "-", "-", nullptr};
    LineDef line = makeDoorLineDef(front, back, v1, v2, fSide, bSide);

    std::vector<std::shared_ptr<SectorMover>> active;
    FakePlayer player;
    player.keys.push_back("blue");

    auto makeMover = [&](LineDef*) {
        return std::make_shared<Door>(&back, "Open", 128.0f, 4.0f, std::optional<std::string>("blue"));
    };
    auto isLockedCheck = [&](LineDef*) {
        return std::make_shared<Door>(&back, "Open", 128.0f, 4.0f, std::optional<std::string>("blue"));
    };

    auto mover = triggerLineSpecial(&line, "Use", &player, active, makeMover, isLockedCheck);
    CHECK(mover != nullptr);
    CHECK_EQ(active.size(), 1u);
}

TEST(advance_specials_removes_finished_keeps_active) {
    Sector sectorA{0, 0, "F", "C", 160, 0, 0}, sectorB{0, 0, "F", "C", 160, 0, 0};
    auto finishedSoon = std::make_shared<FloorMover>(&sectorA, 1.0f, 1.0f);
    auto longRunning = std::make_shared<FloorMover>(&sectorB, 100.0f, 1.0f);
    std::vector<std::shared_ptr<SectorMover>> active = {finishedSoon, longRunning};
    active = advanceSpecials(active);
    CHECK(std::find(active.begin(), active.end(), longRunning) != active.end());
    CHECK(std::find(active.begin(), active.end(), finishedSoon) == active.end());
}

TEST_MAIN()
